File: app/src/main/java/com/hunihun/kotlintest/ScopeFunction.java
package com.hunihun.kotlintest;

import android.os.Build;

import androidx.annotation.RequiresApi;

import static java.lang.Integer.sum;
import static kotlin.random.RandomKt.Random;

public class ScopeFunction {
    private void main() {

    }

    void apply() {
        // Java
        Person person = new Person();
        person.firstName = "홍";
        person.lastName = "길동";
    }

    void also() {
        int value = Random(100).nextInt(100);
        System.out.print(value);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    void let() {
        Integer number = null;
        String sumNumberStr = null;

        if (number != null) {
            sumNumberStr = "" + sum(10, number);
        }
    }

    void with() {
        Person person = new Person();
        person.work();
        person.sleep();
        System.out.println(person.getAge());
    }

    void run() {

    }
}

File: app/src/main/java/com/hunihun/kotlintest/ScopeFunction.kt
package com.hunihun.kotlintest

import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import java.lang.Integer.sum
import kotlin.random.Random

fun main() {
    println()
}

fun apply() {
    val person = Person().apply {
        firstName = "홍"
        lastName = "길동"
    }
}

fun also() {
    Random.nextInt(100).also {
        Log.d("jsh","Random value: $it")
    }

    Random.nextInt(100).also { value ->
        Log.d("jsh","Random value: $value")
    }
}

@RequiresApi(Build.VERSION_CODES.N)
fun let() {
    val number: Int?

    val sumNumberStr = number?.let {
        "${sum(10, it)}"
    }
}

fun with() {
    val person = Person()
    with(person) {
        work()
        sleep()
        Log.d("jsh", age)
    }
}

fun run() {
    val age = Person().run {
        age = "30"
        sleep()
    }
}